CMakeLists.txt

```diff
@@ -11,28 +11,36 @@ project(greentea
 
 include(GNUInstallDirs)
 
-add_library(client)
-target_include_directories(client
+add_library(client_userio
+    source/greentea_test_env.cpp
+)
+target_include_directories(client_userio
     PUBLIC
         "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
         "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>"
 )
 
-target_sources(client
-    PRIVATE
-        source/greentea_test_env.cpp
+add_library(client
+    source/greentea_test_env.cpp
+    source/greentea_test_io.c
+)
+target_include_directories(client
+    PUBLIC
+        "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
```

Comment:
This duplication of include directories and source files isn't ideal. When I tried to inherit the "client_userio" target it caused errors in the CI (but not locally on my macbook using clang or gcc10 or 11) due to missing definitions for the greentea_putc and greentea_getc IO functions. I suspect it was due to link order which somehow the mac versions of GCC and clang were masking.

If someone can suggest a better solution for this, which doesn't cause the link errors, it would be useful.

Reply:
How did you inherit it? Linking one to the other?

Reply:
Yea I can reproduce it on Ubuntu...

Reply:
I guess you did something like
```cmake
add_library(client
    source/greentea_test_io.c
)

target_link_libraries(client
    PUBLIC
        client_userio
)
```
?

What happens is, `libclient.a` will only contain greentea_test_io without any contents of `libclient_userio.a`. Instead, when you build an example, CMake will link both both static libraries to the executable, probably something like `-lclient -lclient_userio`

The linker first links `libclient.a` and decides there's nothing useful, because main() doesn't call any of the the IO function. Then when it proceeds to link `libclient_userio.a`, it can't find the IO functions.

Reply:
In terms of API offering, linking `client` to `client_userio` makes sense. But in terms code dependency, it doesn't. The IO functions in the former are actually called by the latter, so it's a dependency reversal in a sense.

Maybe we could explore a way to fuse `libclient_userio.a` into `libclient.a`, so `libclient.a` will become a self-contained, shippable library with the whole greentea-client instead of just the IO part.

Reply:
Yeah, weirdly with GCC 10 and GCC 11 on my macbook it also linked without issue. It's only when I ran it in the CI, then tried it locally on ArchLinux where the link order became a problem. I'll try and find a way to reverse the dependencies as you suggested.

Reply:
I couldn't really think of a nicer way of doing this without resorting to some INTERFACE library hacking or using variables. I'm fine with leaving it how it is I think. It's a small amount of duplication and in some ways it's probably easier to grok for someone new to CMake.

If there's an actual issue with this then I can spend a bit more time on it. I think I've probably used up enough time on this now though!

Reply:
I'm also fine with it. Another idea is to make the IO functions weak function, but I roughtly remember we would avoid it in the future, possibly to avoid uncertainty in what gets linked.

Reply:
I don't think weak symbols will help, given they get hard resolved before going into a `.a` file. As I understand it, you can't override a symbol (even if you put "weak" in the C file) in a `.a` file with a stronger symbol from another `.a` file: because after the `.a` is make, all symbols are "strong".